hpx/lcos/detail/future_data.hpp

```diff
@@ -219,37 +240,98 @@ namespace detail
+                            std::ref(ptr)),
+                        ec))
+                {
+                    // thread creation went wrong
+                    if (ec) {
+                        boost::unique_lock<mutex_type> l(this->mtx_);
+                        set_result_locked(std::move(l),
+                            hpx::detail::access_exception(ec));
+                        return;
+                    }
+
+                    // re-throw exception in this context
+                    HPX_ASSERT(ptr);        // exception should have been set
+                    boost::rethrow_exception(ptr);
+                }
+            }
+        }
+
         /// Set the result of the requested action.
         template <typename Target>
         void set_result(Target && data, error_code& ec = throws)
         {
-            completed_callback_type on_completed;
-            {
-                boost::unique_lock<mutex_type> l(this->mtx_);
+            boost::unique_lock<mutex_type> l(this->mtx_);
 
-                // check whether the data already has been set
-                if (is_ready_locked()) {
-                    HPX_THROWS_IF(ec, promise_already_satisfied,
-                        "future_data::set_result",
-                        "data has already been set for this future");
-                    return;
-                }
+            // check whether the data has already been set
+            if (is_ready_locked()) {
+                HPX_THROWS_IF(ec, promise_already_satisfied,
+                    "future_data::set_result",
+                    "data has already been set for this future");
+                return;
+            }
+
+            return set_result_locked(std::move(l),
+                std::forward<Target>(data), ec);
+        }
+
+        // Set result unconditionally
+        template <typename Mutex, typename Target>
+        void set_result_locked(boost::unique_lock<Mutex> l, Target && data,
+            error_code& ec = throws)
+        {
+            completed_callback_type on_completed;
 
-                on_completed = std::move(this->on_completed_);
+            on_completed = std::move(this->on_completed_);
 
-                // set the data
-                data_ = std::forward<Target>(data);
+            // set the data
+            data_ = std::forward<Target>(data);
 
-                // make sure the entry is full
-                state_ = full;
+            // make sure the entry is full
+            state_ = full;
 
-                // handle all threads waiting for the block to become full
-                cond_.notify_all(std::move(l), ec);
-            }
+            // handle all threads waiting for the future to become ready
+            cond_.notify_all(std::move(l), ec);
 
             // invoke the callback (continuation) function
             if (on_completed)
-                on_completed();
+                handle_on_completed(std::move(on_completed));
```

Comment:
Yes please, this is a very odd looking code construct. But seems that it works as intended.